File: app/src/main/java/com/blogspot/e_kanivets/moneytracker/controller/external/ImportController.java
package com.blogspot.e_kanivets.moneytracker.controller.external;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.blogspot.e_kanivets.moneytracker.controller.data.RecordController;
import com.blogspot.e_kanivets.moneytracker.entity.data.Account;
import com.blogspot.e_kanivets.moneytracker.entity.data.Category;
import com.blogspot.e_kanivets.moneytracker.entity.data.Record;

import java.util.ArrayList;
import java.util.List;

/**
 * Controller class to encapsulate import logic.
 * Created on 6/28/16.
 *
 * @author Evgenii Kanivets
 */
public class ImportController {
    private final RecordController recordController;

    public ImportController(RecordController recordController) {
        this.recordController = recordController;
    }

    @NonNull
    public List<Record> importRecordsFromCsv(@Nullable String csv) {
        List<Record> recordList = new ArrayList<>();
        if (csv == null) return recordList;

        String lines[] = csv.split("\\r?\\n");
        for (String line : lines) {
            String[] words = line.split(Head.DELIMITER);
            if (words.length != Head.COLUMN_COUNT) continue;

            String timeCol = words[0];
            String accountIdCol = words[1];
            String titleCol = words[2];
            String categoryCol = words[3];
            String notesCol = words[4];
            String priceCol = words[5];
            String currencyCol = words[6];


            try {
                long time = Long.parseLong(timeCol);
                long accountId = Long.parseLong(accountIdCol);
                String title = titleCol.trim();
                String categoryName = categoryCol.trim();
                String notes = notesCol.trim();
                double price = Double.parseDouble(priceCol);
                String currency = currencyCol.trim();

                if (currency.length() != 3) continue;
                if (categoryName.isEmpty()) continue;

                int type;
                if (price < 0.0) type = Record.TYPE_EXPENSE;
                else type = Record.TYPE_INCOME;

                Category category = new Category(categoryName);
                Account account = new Account(accountId, "MOCK", -1, currency, 0, -1, false, 0);

                Record record = new Record(time, type, title, category, notes, Math.abs(price), account, currency);
                Record createdRecord = recordController.create(record);
                if (createdRecord != null) recordList.add(createdRecord);
            } catch (Exception e) {
                e.printStackTrace();
            }
        }

        return recordList;
    }
}

File: app/src/main/java/com/blogspot/e_kanivets/moneytracker/controller/external/ImportController.kt
package com.blogspot.e_kanivets.moneytracker.controller.external

import com.blogspot.e_kanivets.moneytracker.controller.data.RecordController
import com.blogspot.e_kanivets.moneytracker.entity.data.Account
import com.blogspot.e_kanivets.moneytracker.entity.data.Category
import com.blogspot.e_kanivets.moneytracker.entity.data.Record

/**
 * Controller class to encapsulate import logic.
 * Created on 6/28/16.
 *
 * @author Evgenii Kanivets
 */
class ImportController(private val recordController: RecordController) {

    fun importRecordsFromCsv(csv: String?): List<Record> {
        val recordList = mutableListOf<Record>()
        if (csv == null) return recordList

        val lines = csv.split("\\r?\\n".toRegex())
        for (line in lines) {
            val words = line.split(Head.DELIMITER)
            if (words.size != Head.COLUMN_COUNT) continue

            val timeCol = words[0]
            val accountIdCol = words[1]
            val titleCol = words[2]
            val categoryCol = words[3]
            val notesCol = words[4]
            val priceCol = words[5]
            val currencyCol = words[6]

            try {
                val time = timeCol.toLong()
                val accountId = accountIdCol.toLong()
                val title = titleCol.trim()
                val categoryName = categoryCol.trim()
                val notes = notesCol.trim()
                val price = priceCol.toDouble()
                val currency = currencyCol.trim()

                if (currency.length != 3) continue
                if (categoryName.isEmpty()) continue

                val type = if (price < 0.0) Record.TYPE_EXPENSE else Record.TYPE_INCOME

                val category = Category(categoryName)
                val account = Account(accountId, "MOCK", -1, currency, 0, -1.0, false, 0)

                val record = Record(time, type, title, category, notes, Math.abs(price), account, currency)
                val createdRecord = recordController.create(record)
                if (createdRecord != null) recordList.add(createdRecord)
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }

        return recordList
    }
}
